File: src/BackwardDataflow.fs
module Backward_dataflow

type Annotation<'varSet> = 'varSet
type AnnotatedBlock<'a, 'block> = Cfg.BasicBlock<'a, 'block>
type AnnotatedGraph<'a, 'block> = Cfg.ControlFlowGraph<'a, 'block>

let debugPrint (debug: bool) (extra_tag: string) (pp_var: System.IO.TextWriter -> 'var -> unit)
                (elements: 'varSet -> 'var list)
                (print_graphviz: (System.IO.TextWriter -> 'varSet -> unit) -> 'cfg -> unit)
                (debugLabel: string)
                (setDebugLabel: string -> 'cfg -> 'cfg)
                (cfg: 'cfg) =
    if debug then
        let livevarPrinter (fmt: System.IO.TextWriter) (liveVars: 'varSet) =
            elements liveVars
            |> List.iteri (fun i v ->
                if i > 0 then fmt.Write(", ")
                pp_var fmt v)
        let lbl = debugLabel + "_dse" + extra_tag
        print_graphviz livevarPrinter (setDebugLabel lbl cfg)

let analyze (debug: bool) (pp_var: System.IO.TextWriter -> 'var -> unit)
            (empty: 'varSet)
            (equal: 'varSet -> 'varSet -> bool)
            (elements: 'varSet -> 'var list)
            (meet_fn: 'cfg -> 'block -> 'varSet)
            (transfer_fn: 'block -> 'varSet -> 'block)
            (initialize_annotation: 'cfg0 -> 'varSet -> 'cfg)
            (update_basic_block: int -> 'block -> 'cfg -> 'cfg)
            (get_value: 'block -> 'varSet)
            (get_preds: 'block -> Cfg.NodeId list)
            (get_basic_blocks: 'cfg -> (int * 'block) list)
            (getDebugLabel: 'cfg -> string)
            (setDebugLabel: string -> 'cfg -> 'cfg)
            (print_graphviz: (System.IO.TextWriter -> 'varSet -> unit) -> 'cfg -> unit)
            (cfg: 'cfg0) =
    let startingCfg = initialize_annotation cfg empty
    let rec processWorklist (currentCfg: 'cfg)
                             (worklist: (int * 'block) list) =
        debugPrint debug "_in_progress_" pp_var elements print_graphviz
            (getDebugLabel currentCfg) setDebugLabel currentCfg
        match worklist with
        | [] -> currentCfg // we're done
        | (blockIdx, blk) :: rest ->
            let oldAnnotation = get_value blk
            let liveVarsAtExit = meet_fn currentCfg blk
            let block' = transfer_fn blk liveVarsAtExit
            let updatedCfg = update_basic_block blockIdx block' currentCfg
            let newWorklist =
                if equal oldAnnotation (get_value block') then rest
                else
                    List.fold
                        (fun wklist pred ->
                            match pred with
                            | Cfg.Entry -> wklist
                            | Cfg.Exit ->
                                failwith "Internal error: malformed CFG"
                            | Cfg.Block n ->
                                if List.exists (fun (k, _) -> k = n) wklist then wklist
                                else
                                    let blocks = get_basic_blocks updatedCfg
                                    let blk =
                                        match List.tryFind (fun (k, _) -> k = n) blocks with
                                        | Some (_, b) -> b
                                        | None -> failwith "Internal error: block not found in CFG"
                                    (n, blk) :: wklist)
                        rest (get_preds block')
            processWorklist updatedCfg newWorklist
    processWorklist startingCfg (List.rev (get_basic_blocks startingCfg))

File: src/BackwardDataflow.rs
//! Generic work-list solver for backwards data-flow analyses.
#![allow(non_snake_case)]

use std::io::{self, Write};

use super::Cfg::{BasicBlock, ControlFlowGraph, NodeId};

pub type Annotation<VarSet> = VarSet;
pub type AnnotatedBlock<VarSet, Instruction> = BasicBlock<VarSet, Instruction>;
pub type AnnotatedGraph<VarSet, Instruction> = ControlFlowGraph<VarSet, Instruction>;

#[allow(clippy::too_many_arguments)]
pub fn debugPrint<Var, VarSet, Cfg, PP, Elements, Print, Label, SetLabel>(
    debug: bool,
    extra_tag: &str,
    pp_var: &PP,
    elements: &Elements,
    print_graphviz: &Print,
    debug_label: &Label,
    set_debug_label: &SetLabel,
    cfg: &Cfg,
) -> io::Result<()>
where
    PP: Fn(&mut dyn Write, &Var) -> io::Result<()>,
    Elements: Fn(&VarSet) -> Vec<Var>,
    Print: Fn(&mut dyn FnMut(&mut dyn Write, &VarSet) -> io::Result<()>, &Cfg) -> io::Result<()>,
    Label: Fn(&Cfg) -> &str,
    SetLabel: Fn(String, &Cfg) -> Cfg,
{
    if debug {
        let mut livevar_printer = |out: &mut dyn Write, live_vars: &VarSet| {
            for (index, variable) in elements(live_vars).iter().enumerate() {
                if index > 0 { write!(out, ", ")?; }
                pp_var(out, variable)?;
            }
            Ok(())
        };
        let label = format!("{}_dse{}", debug_label(cfg), extra_tag);
        print_graphviz(&mut livevar_printer, &set_debug_label(label, cfg))?;
    }
    Ok(())
}

/// Run a backwards analysis to a fixed point.
///
/// As in the F# implementation, blocks initially enter the work list in
/// reverse CFG order and a changed block schedules each unscheduled predecessor
/// at the front of the list.
#[allow(clippy::too_many_arguments)]
pub fn analyze<Var, VarSet, Block, Cfg0, Cfg, PP, Elements, Meet, Transfer, Initialize,
    Update, Value, Preds, Blocks, Label, SetLabel, Print>(
    debug: bool, pp_var: PP, empty: VarSet, equal: impl Fn(&VarSet, &VarSet) -> bool,
    elements: Elements, meet_fn: Meet, transfer_fn: Transfer,
    initialize_annotation: Initialize, update_basic_block: Update, get_value: Value,
    get_preds: Preds, get_basic_blocks: Blocks, get_debug_label: Label,
    set_debug_label: SetLabel, print_graphviz: Print, cfg: Cfg0,
) -> io::Result<Cfg>
where
    VarSet: Clone,
    Block: Clone,
    PP: Fn(&mut dyn Write, &Var) -> io::Result<()>,
    Elements: Fn(&VarSet) -> Vec<Var>,
    Meet: Fn(&Cfg, &Block) -> VarSet,
    Transfer: Fn(Block, VarSet) -> Block,
    Initialize: Fn(Cfg0, VarSet) -> Cfg,
    Update: Fn(usize, Block, &mut Cfg),
    Value: Fn(&Block) -> &VarSet,
    Preds: Fn(&Block) -> &[NodeId],
    Blocks: Fn(&Cfg) -> Vec<(usize, Block)>,
    Label: Fn(&Cfg) -> &str,
    SetLabel: Fn(String, &Cfg) -> Cfg,
    Print: Fn(&mut dyn FnMut(&mut dyn Write, &VarSet) -> io::Result<()>, &Cfg) -> io::Result<()>,
{
    let mut current_cfg = initialize_annotation(cfg, empty);
    let mut worklist = get_basic_blocks(&current_cfg);
    worklist.reverse();
    loop {
        debugPrint(debug, "_in_progress_", &pp_var, &elements, &print_graphviz,
                   &get_debug_label, &set_debug_label, &current_cfg)?;
        let Some((block_index, block)) = worklist.first().cloned() else { return Ok(current_cfg); };
        worklist.remove(0);
        let old_annotation = get_value(&block).clone();
        let live_vars_at_exit = meet_fn(&current_cfg, &block);
        let updated_block = transfer_fn(block, live_vars_at_exit);
        let changed = !equal(&old_annotation, get_value(&updated_block));
        let predecessors = get_preds(&updated_block).to_vec();
        update_basic_block(block_index, updated_block, &mut current_cfg);
        if changed {
            for predecessor in predecessors {
                match predecessor {
                    NodeId::Entry => {}
                    NodeId::Exit => panic!("Internal error: malformed CFG"),
                    NodeId::Block(number) if !worklist.iter().any(|(index, _)| *index == number) => {
                        let predecessor_block = get_basic_blocks(&current_cfg).into_iter()
                            .find(|(index, _)| *index == number).map(|(_, block)| block)
                            .expect("Internal error: block not found in CFG");
                        worklist.insert(0, (number, predecessor_block));
                    }
                    NodeId::Block(_) => {}
                }
            }
        }
    }
}
